internal/migrations/changeset_listener.go

```diff
@@ -0,0 +1,272 @@
+	if err != nil {
+		panic(err)
+	}
+}
+
+type MigrationConfig struct {
+	// StartHeight is the block height at which the migration started on the old chain.
+	StartHeight int64
+
+	// End height is the block height at which the migration ended on the old chain.
+	EndHeight int64
+
+	// AdminListenAddress is the address of the admin server to receive changesets from.
+	AdminListenAddress string
+
+	// AdminPass is the admin server password.
+	AdminPass string
+
+	// KwildTLSCertFile is the path to the TLS certificate file for the Kwil node.
+	// The path should be an absolute path or relative to the node's root directory.
+	KwildTLSCertFile string
+}
+
+func Start(ctx context.Context, service *common.Service, eventStore listeners.EventStore) error {
+	// Get the migration config from the service
+	cfg := &MigrationConfig{}
+	migrationCfg, ok := service.ExtensionConfigs[ListenerName]
+	if !ok {
+		service.Logger.Warn("no migration config provided, skipping migration listener")
+		return nil // no migration config, nothing to do
+	}
+
+	err := cfg.ExtractConfig(migrationCfg)
+	if err != nil {
+		return err
+	}
+
+	var dialOpt []adminclient.Opt
+	if cfg.KwildTLSCertFile != "" {
+		dialOpt = append(dialOpt, adminclient.WithTLS(cfg.KwildTLSCertFile, "", ""))
+	}
+
+	if cfg.AdminPass != "" {
+		dialOpt = append(dialOpt, adminclient.WithPass(cfg.AdminPass))
+	}
+
+	// Admin RPC client connection to the old chain
+	adminClient, err := adminclient.NewClient(ctx, cfg.AdminListenAddress, dialOpt...)
+	if err != nil {
+		return err
+	}
+
+	// Get the block height of the last changesets received from the old chain
+	// If no height is found, start from the start height of the migrations.
+	// If all the changesets have already been received, the listener will exit.
+	var currentHeight, lastHeight int64
+	lastHeight, err = getLastStoredHeight(ctx, eventStore)
+	if err != nil {
+		return err
+	}
+
+	if lastHeight == 0 {
+		currentHeight = cfg.StartHeight
+	} else if lastHeight >= cfg.EndHeight {
+		service.Logger.Info("all the changesets have been synced to the new chain, closing the migrations listener.", log.Int("height", lastHeight))
+		return nil
+	} else {
+		currentHeight = lastHeight + 1
+	}
+
+	// Start polling the admin server for changesets
+	service.Logger.Info("start syncing changesets from old chain", log.Int("startHeight", cfg.StartHeight))
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			if currentHeight >= cfg.EndHeight {
+				// Synced up with the old chain, no more changesets to pull. Close the listener
+				service.Logger.Info("changesets have been synchronized with the old chain", log.Int("height", currentHeight))
+				return nil
+			}
+
+			// Get the changeset metadata from the admin server
+			numChunks, size, err := adminClient.ChangesetMetadata(ctx, currentHeight)
+			if err != nil {
+				fmt.Println("Couldnt fetch changesets", err)
+				time.Sleep(retryDelay) // If no changeset is found, wait and try again
```

Comment:
Should try to avoid `time.Sleep` in almost all cases, here and in the goroutines below.  The main reason is shutdown can't bypass, as it's stuck waiting even if context canceled.   This will be especially irritating if you move to an exponential backoff as the delays will just get longer.  Better to refactor with a receive from a timer/ticker/After with a competing case on the context in case of shutdown.